Lib/Database/MySqlQuery.php

```diff
@@ -0,0 +1,1247 @@
+   * string.
+   */
+  public function whereNotIn($field, array $values) {
+    try {
+      return $this->processWhereIn($field, $values, 'whereNotIn');
+    } catch (Exception $e) {
+      throw $e;
+    }
+  }
+
+  /**
+   * Check query result existence.
+   *
+   * @since 1.0.0
+   * @access public
+   *
+   * @return boolean Whether the result is set or not.
+   */
+  public function hasResult() {
+    return isset($this->result);
+  }
+
+  /**
+   * Set query result.
+   *
+   * @since 1.0.0
+   * @access public
+   *
+   * @param mixed $result Query result.
+   */
+  public function setResult($result) {
+    $this->result = $result;
+  }
+
+  /**
+   * Retrieve query result.
+   *
+   * Retrive result(s) of the query. If result isn't set, null will be returned.
+   *
+   * @since 1.0.0
+   * @access public
+   *
+   * @return mixed Query result.
+   */
+  public function getResult() {
+    return $this->result;
+  }
+
+  /**
+   * Build the query string.
+   *
+   * @since 1.0.0
+   * @access public
+   *
+   * @return string
+   *
+   * @throws \RuntimeException if $table property isn't set or if there is some
+   * inconsistency with the data required by every specific method.
+   */
+  public function build() {
+    try {
+      switch ($this->type) {
+        case 'select':
+          return $this->buildSelect();
+          break;
+        case 'update':
+          return $this->buildUpdate();
+          break;
+        case 'insert':
+          return $this->buildInsert();
+          break;
+        case 'delete':
+          return $this->buildDelete();
+          break;
+      }
+    } catch (Exception $e) {
+      throw $e;
+    }
+  }
+
+  /**
+   * Process WHERE clauses.
+   *
+   * @since 1.0.0
+   * @access private
+   *
+   * @param array $args Arguments of the public method.
+   * @param string $method Name of the method that called this one.
+   * @return self
+   *
+   * @throws \RuntimeException if the arguments aren't in the correct form.
+   */
+  private function processWhere(array $args, $method = 'where') {
+    if (is_array($args[0])) {
+      // Method was called with an array of clauses.
+      foreach ($args[0] as $clause) {
+        // Check if the array has other arrays or not, and if it has , then
+        // check if every internal array has only scalar inside.
+        if (!is_array($clause) || count($clause) < 2 || count($clause) > 3)
+          throw new RuntimeException('Arguments passed to `Query::'.$method.'()` aren\'t in the correct form');
+        $count = 0;
+        foreach ($clause as $scalar) {
+          if (!is_scalar($scalar) && (!is_array($scalar) || $count != 3))
+            throw new RuntimeException('Arguments passed to `Query::'.$method.'()` aren\'t in the correct form');
+          $count++;
+        }
+      }
+
+      // The array is in the correct form.
+      try {
+        $this->buildWhereClause($args[0], $method == 'where' ? 'AND' : 'OR');
```

Comment:
It's easier if you pass the correct `$method` (`AND`, `OR`) to the method directly.